system_tests/openstack_handler.py

```diff
@@ -43,24 +46,71 @@ def __init__(self, context_name, env):
-        resources_to_teardown = self.get_resources_to_teardown()
+        resources_to_teardown = self.get_resources_to_teardown(
+            self.env, resources_to_keep=self.before_run)
+
         if self.skip_cleanup:
             self.logger.warn('[{0}] SKIPPING cleanup: of the resources: {1}'
                              .format(self.context_name, resources_to_teardown))
             return
+
         self.logger.info('[{0}] Performing cleanup: will try removing these '
                          'resources: {1}'
                          .format(self.context_name, resources_to_teardown))
-
         leftovers = self.env.handler.remove_openstack_resources(
             resources_to_teardown)
         self.logger.info('[{0}] Leftover resources after cleanup: {1}'
                          .format(self.context_name, leftovers))
 
-    def get_resources_to_teardown(self):
-        current_state = self.env.handler.openstack_infra_state()
-        return self.env.handler.openstack_infra_state_delta(
-            before=self.before_run, after=current_state)
+    @classmethod
+    def clean_all(cls, env):
+        super(OpenstackCleanupContext, cls).clean_all(env)
+        resources_to_teardown = cls.get_resources_to_teardown(env)
+        cls.logger.info('Openstack handler performing clean_all: will try '
+                        'removing these resources: {0}'
+                        .format(resources_to_teardown))
+        failed_to_remove = env.handler.remove_openstack_resources(
+            resources_to_teardown)
+        if failed_to_remove:
+            trimmed_dict = {key: value for key, value in
+                            failed_to_remove.iteritems()
+                            if value}
+            if len(trimmed_dict) > 0:
+                msg = 'Openstack handler failed to remove some resources' \
+                      ' during clean_all: {0}'.format(trimmed_dict)
+                cls.logger.error(msg)
+                raise RuntimeError(msg)
+
+    @classmethod
+    def get_resources_to_teardown(cls, env, resources_to_keep=None):
+        existing_resources = env.handler.openstack_infra_state()
+        return cls.filter_out_resources(
+            existing_resources, resources_to_keep)
+
+    @classmethod
+    def filter_out_resources(cls, resources_by_type, resources_to_filter):
+        if not resources_to_filter:
+            return resources_by_type
+
+        for resource_type, resources_to_filter_of_type in \
+                resources_to_filter.iteritems():
+            existing_resources_of_type = resources_by_type.get(resource_type)
+            if existing_resources_of_type:
+                # remove by key (resource id) or by value (resource name)
+                for resource_id, resource_name in \
+                        resources_to_filter_of_type.iteritems():
+                    if resource_id and resource_id in \
+                            existing_resources_of_type.keys():
```

Comment:
no need for `.keys()`, this is implicit if you only do `resource_id in existing_resources_of_type`